Mathlib/NumberTheory/SiegelsLemma.lean

```diff
@@ -0,0 +1,242 @@
+  simp only [ne_eq, Nat.cast_inj]
+  linarith only [hn]
+
+-- (2)
+
+private lemma card_S_lt_card_T : (S).card < (T).card := by
+  zify
+  rw [card_T_eq, card_S_eq]
+  let a := natNorm A
+  calc
+  ∏ i : Fin m, (P i - N i + 1) ≤ (n * a * B + 1) ^ m := by
+        rw [← Fin.prod_const m ((n * a * B + 1) : ℤ)]
+        apply Finset.prod_le_prod  --2 goals
+        all_goals intro i _
+        linarith only [N_le_P_add_one m n A i] --first goal done
+        simp only [mul_neg, sum_neg_distrib, sub_neg_eq_add, cast_succ, cast_mul,
+          add_le_add_iff_right]
+        rw [(mul_sum Finset.univ (fun i_1 => (A i i_1)⁺) ↑B).symm,
+            (mul_sum Finset.univ (fun i_1 => (A i i_1)⁻) ↑B).symm, ← mul_add,
+            mul_comm ((n : ℤ) * a) (B : ℤ)]
+        apply mul_le_mul_of_nonneg_left _ (cast_nonneg B)
+        rw [← Finset.sum_add_distrib]
+        have h1 : n * (a : ℤ) = ∑ _ : Fin n, (a : ℤ) := by
+            simp only [sum_const, card_fin, nsmul_eq_mul]
+        rw [h1]
+        gcongr with j _
+        rw [posPart_add_negPart (A i j)]
+        have h2 : |A i j| ≤ (a : ℝ) := by
+                  rw [Int.cast_abs, ← Int.norm_eq_abs, ← norm_eq_natNorm]
+                  exact norm_entry_le_entrywise_sup_norm A
+        exact Int.cast_le.1 h2
+      _  ≤ (n * a) ^ m * (B + 1) ^ m := by
+        rw [← mul_pow (n * (a : ℤ)) ((B : ℤ) + 1) m]
+        apply pow_le_pow_left (Int.ofNat_nonneg (n*a*B+1))
+        rw [mul_add]
+        simp only [cast_succ, cast_mul, mul_one, add_le_add_iff_left]
+        exact_mod_cast one_le_mul (one_le_of_lt hn) (one_le_natNorm_of_ne_zero A hA)
+      _ < (B + 1) ^ (n - m) * (B + 1) ^ m := by
+        simp only [gt_iff_lt, Int.succ_ofNat_pos, pow_pos, mul_lt_mul_right]
+        have h1 : ((n * a) ^ (m / ((n : ℝ) - m))) ^ ((n : ℝ) - m) <
+            ((B + 1): ℝ) ^ ((n : ℝ) - m) := by -- prove the goal with rpow
+          apply Real.rpow_lt_rpow /- this creates 3 goals: 0 ≤ (↑n * ↑a) ^ (↑m / (↑n - ↑m)),
+                                      (↑n * ↑a) ^ (↑m / (↑n - ↑m)) < ↑B + 1 and 0 < ↑n - ↑m -/
+          · apply rpow_nonneg
+            exact_mod_cast Nat.zero_le (n * a)
+          · rw [← norm_eq_natNorm]
+            exact lt_floor_add_one ((n * ‖A‖) ^ (m / ((n : ℝ ) - ↑m)))
+          · simp only [sub_pos, cast_lt, hn]
+        have h2 : ((n * a) ^ (m / ((n : ℝ) - m))) ^ ((n : ℝ) - m) =
+            ((n : ℝ) * a) ^ (m : ℝ) := by
+          rw [← rpow_mul (by exact_mod_cast Nat.zero_le (n * a)) (m / (n - m)) (n - m),
+            hcompexp m n hn]
+        rw [h2] at h1
+        nth_rw 2 [← Nat.cast_sub (le_of_lt hn)] at h1
+        rw [rpow_natCast ((↑B + 1)) (n - m), rpow_natCast ] at h1
+        exact_mod_cast h1
+      _  = ↑((B + 1) ^ n) := by
+        rw [mul_comm, pow_mul_pow_sub]
+        simp only [cast_pow, cast_add, cast_one]
+        exact (le_of_lt hn)
+
+--end (2)
+
+end preparation
+
+theorem exists_ne_zero_int_vec_norm_le  (hA_nezero : A ≠ 0)  : ∃ (t : Fin n → ℤ), t ≠ 0 ∧
```

Comment:
We are in the root namespace, right? I guess `Matrix` or something would be better.